### PR TITLE
[IMP] website_blog: prefetch the first post with the other posts

### PR DESCRIPTION
**Pull request by seb-odoo**
This reduces the number of queries by 4 on this route.

Here are the solutions that have been considered on how to share this prefetch.

[0].name
--------

Calling `(first_post | posts)[0].name` does not work because `[0]` is another
prefetch, it still does all the queries.

See `__getitem__` in `models.py`,  `_prefetch_ids` is not forwarded, as opposed
to `__iter__`.

.mapped('name')
---------------

Calling `(first_post | posts).mapped('name')` reduces the number of queries by 2
(because it does an iteration = share prefetch, but this is CPU time for
nothing).

The problem is that even in that case, when computing the display name of a blog
for the slug (`slug(blog_post.blog_id)` or `slug(first_post.blog_id)`) if the
posts are on a different prefetch, the query will be done twice because their
m2o (blog_id in this case) will be on a different prefetch each.

.blog_id.mapped('name')
-----------------------

Calling `(first_post | posts).blog_id.mapped('name')`. would be the same as
using `with_prefetch` in this case, but it will be less robust. It is also
slightly more CPU to do the map for no reason, though it is really not a factor.

But the main problem is that it only works because we know `.blog_id.name` will
be accessed, but if the view is changed some day to access a field of another
m2o, this will not be optimized, while `with_prefetch` it is always guaranteed.

In conclusion the solution with `with_prefetch` was used because it is explicit,
more robust, and without any side effect.

Part of task-2061122

### CODE REVIEW

**Comment by seb-odoo** [comment]
[0].name
-----------

As discussed I tested with `(first_post | posts)[0].name` but it does not work because `[0]` is another prefetch, it still does all the queries. 

If we have that in our code for prefetch it is not working and should be adapted.

See `__getitem__` in `models.py`,  `_prefetch_ids` is not forwarded, as opposted to `__iter__`.

This might be a flaw of the ORM.
- This would be OK if when accessing by key, we want only that record and ignore the others (and then why search multiple records in the first place, it is a waste).
- And if multiple records are accessed by key, or if one is accessed by key and then the full recordset is accessed too later, then it is bad that prefetch is not shared.

So it seems always bad to not share prefetch on `__getitem__`.

.mapped('name')
---------------------

Then I tested with `(first_post | posts).mapped('name')` which reduces the number of queries by 2 (because it does an iteration = share prefetch, but this is CPU time for nothing).

The problem is that even in that case, when computing the display name of a blog for the slug (`slug(blog_post.blog_id)` or `slug(first_post.blog_id)`) if the posts are on a different prefetch, the query will be done twice because their m2o (blog_id in this case) will be on a different prefetch each.

.blog_id.mapped('name')
-------------------------------

There is one solution, it is to call `(first_post | posts).blog_id.mapped('name')`.
Then it is the same as using `with_prefetch` in this case, but it will be less robust:
- it is also slightly more CPU to do the map for no reason, though it is really not a factor
- but the main problem is that it only works because we know `.blog_id.name` will be accessed, but if the view is changed some day to access a field of another m2o, this will not be optimized, while `with_prefetch` it is guaranteed always ok.


So in conclusion I would keep `with_prefetch` because it is explicit, more robust, and without any side effect.

**Comment by JKE-be** [comment]
@robodoo r+

Good to know, thx
